File: myExpenses/src/main/java/org/totschnig/myexpenses/util/licence/PlayStoreLicenceHandler.java
package org.totschnig.myexpenses.util.licence;

import android.app.Activity;
import android.content.SharedPreferences;
import android.text.TextUtils;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsResponseListener;
import com.annimon.stream.Stream;
import com.google.android.vending.licensing.PreferenceObfuscator;

import org.json.JSONException;
import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.activity.ContribInfoDialogActivity;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.util.crashreporting.CrashHandler;

import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

public class PlayStoreLicenceHandler extends AbstractInAppPurchaseLicenceHandler {
  public PlayStoreLicenceHandler(MyApplication context, PreferenceObfuscator preferenceObfuscator, CrashHandler crashHandler) {
    super(context, preferenceObfuscator, crashHandler);
  }

  private void storeSkuDetails(List<SkuDetails> inventory) {
    SharedPreferences.Editor editor = getPricesPrefs().edit();
    for (SkuDetails skuDetails : inventory) {
      log().w("Sku %s, json: %s", skuDetails.getSku(), skuDetails.toString());
      editor.putString(prefKeyForSkuJson(skuDetails.getSku()), skuDetails.getOriginalJson());
    }
    editor.apply();
  }

  private String prefKeyForSkuJson(String sku) {
    return String.format(Locale.ROOT, "%s_json", sku);
  }

  @Nullable
  protected String getDisplayPriceForPackage(@NonNull Package aPackage) {
    String sku = getSkuForPackage(aPackage);
    SkuDetails skuDetails = getSkuDetailsFromPrefs(sku);
    String result = null;
    if (skuDetails != null) {
      result = skuDetails.getIntroductoryPrice();
      if (TextUtils.isEmpty(result)) {
        result = skuDetails.getPrice();
      }
    }
    return result;
  }

  @Nullable
  private SkuDetails getSkuDetailsFromPrefs(String sku) {
    String originalJson = getPricesPrefs().getString(prefKeyForSkuJson(sku), null);
    if (originalJson != null) {
      try {
        return new SkuDetails(originalJson);
      } catch (JSONException e) {
        CrashHandler.report(e, String.format("unable to parse %s", originalJson));
      }
    } else {
      CrashHandler.report(String.format("originalJson not found for %s", sku));
    }
    return null;
  }

  private String getCurrentSubscription() {
    return licenseStatusPrefs.getString(KEY_CURRENT_SUBSCRIPTION, null);
  }

  @Override
  public BillingManagerPlay initBillingManager(@NonNull Activity activity, boolean query) {
    BillingUpdatesListener billingUpdatesListener = new BillingUpdatesListener() {

      @Override
      public boolean onPurchasesUpdated(@Nullable List<Purchase> inventory) {
        boolean result = false;
        if (inventory != null) {
          LicenceStatus oldStatus = getLicenceStatus();
          result = registerInventory(inventory) != null;
          if (activity instanceof BillingListener) {
            ((BillingListener) activity).onLicenceStatusSet(getLicenceStatus(), oldStatus);
          }
        }
        return result;
      }

      @Override
      public void onPurchaseCanceled() {
        log().i("onPurchasesUpdated() - user cancelled the purchase flow - skipping");
        if (activity instanceof ContribInfoDialogActivity) {
          ((ContribInfoDialogActivity) activity).onPurchaseCancelled();
        }
      }

      @Override
      public void onPurchaseFailed(int resultcode) {
        log().w("onPurchasesUpdated() got unknown resultCode: %s", resultcode);
        if (activity instanceof ContribInfoDialogActivity) {
          ((ContribInfoDialogActivity) activity).onPurchaseFailed(resultcode);
        }
      }
    };
    final SkuDetailsResponseListener skuDetailsResponseListener = query ? (result, skuDetailsList) -> {
      if (result.getResponseCode() == BillingClient.BillingResponseCode.OK && skuDetailsList != null) {
        storeSkuDetails(skuDetailsList);
      } else {
        log().d("skuDetails response %d", result.getResponseCode());
      }
    } : null;
    return new BillingManagerPlay(activity, billingUpdatesListener, skuDetailsResponseListener);
  }

  @VisibleForTesting
  @Nullable
  public Purchase findHighestValidPurchase(List<Purchase> inventory) {
    return Stream.of(inventory)
        .filter(purchase -> extractLicenceStatusFromSku(purchase.getSku()) != null)
        .max((o, o2) -> Utils.compare(extractLicenceStatusFromSku(o.getSku()), extractLicenceStatusFromSku(o2.getSku()), Enum::compareTo))
        .orElse(null);
  }


  private LicenceStatus registerInventory(@NonNull List<Purchase> inventory) {
    Stream.of(inventory).forEach(purchase -> {
      log().i("%s (acknowledged %b)", purchase.getSku(), purchase.isAcknowledged());
    });
    Purchase purchase = findHighestValidPurchase(inventory);
    if (purchase != null) {
      if (purchase.getPurchaseState() == Purchase.PurchaseState.PURCHASED) {
        return handlePurchase(purchase.getSku(), purchase.getOrderId());
      } else {
        CrashHandler.reportWithTag(String.format("Found purchase in state %s", purchase.getPurchaseState()), TAG);
      }
    } else {
      maybeCancel();
    }
    return null;
  }

  @Override
  public void launchPurchase(Package aPackage, boolean shouldReplaceExisting, @NonNull BillingManager billingManager) {
    String sku = getSkuForPackage(aPackage);
    SkuDetails skuDetails = getSkuDetailsFromPrefs(sku);
    if (skuDetails == null) {
      throw new IllegalStateException("Could not determine sku details");
    }
    String oldSku;
    if (shouldReplaceExisting) {
      oldSku = getCurrentSubscription();
      if (oldSku == null) {
        throw new IllegalStateException("Could not determine current subscription");
      }
    } else {
      oldSku = null;
    }
    ((BillingManagerPlay) billingManager).initiatePurchaseFlow(skuDetails, oldSku);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/util/licence/PlayStoreLicenceHandler.kt
package org.totschnig.myexpenses.util.licence

import android.app.Activity
import android.text.TextUtils
import androidx.annotation.VisibleForTesting
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.SkuDetails
import com.android.billingclient.api.SkuDetailsResponseListener
import com.google.android.vending.licensing.PreferenceObfuscator
import org.json.JSONException
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.activity.ContribInfoDialogActivity
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.util.crashreporting.CrashHandler
import java.util.*

open class PlayStoreLicenceHandler(context: MyApplication, preferenceObfuscator: PreferenceObfuscator, crashHandler: CrashHandler, prefHandler: PrefHandler) : AbstractInAppPurchaseLicenceHandler(context, preferenceObfuscator, crashHandler, prefHandler) {
    private fun storeSkuDetails(inventory: List<SkuDetails>) {
        val editor = pricesPrefs.edit()
        for (skuDetails in inventory) {
            log().w("Sku %s, json: %s", skuDetails.sku, skuDetails.toString())
            editor.putString(prefKeyForSkuJson(skuDetails.sku), skuDetails.originalJson)
        }
        editor.apply()
    }

    private fun prefKeyForSkuJson(sku: String): String {
        return String.format(Locale.ROOT, "%s_json", sku)
    }

    override fun getDisplayPriceForPackage(aPackage: Package) =
            getSkuDetailsFromPrefs(getSkuForPackage(aPackage))?.let { skuDetails ->
                skuDetails.introductoryPrice.takeIf { !TextUtils.isEmpty(it) } ?: skuDetails.price
            }

    fun getSkuDetailsFromPrefs(sku: String): SkuDetails? {
        pricesPrefs.getString(prefKeyForSkuJson(sku), null)?.let {
            return try {
                return SkuDetails(it)
            } catch (e: JSONException) {
                CrashHandler.report(e, String.format("unable to parse %s", it))
                null
            }
        } ?: run {
            CrashHandler.report(String.format("originalJson not found for %s", sku))
        }
        return null
    }

    private val currentSubscription: String?
        get() = licenseStatusPrefs.getString(KEY_CURRENT_SUBSCRIPTION, null)

    override fun initBillingManager(activity: Activity, query: Boolean): BillingManager {
        val billingUpdatesListener: BillingUpdatesListener = object : BillingUpdatesListener {
            override fun onPurchasesUpdated(purchases: List<Purchase>?, newPurchase: Boolean): Boolean {
                if (purchases != null) {
                    val oldStatus = licenceStatus
                    val oldFeatures = addOnFeatures
                    registerInventory(purchases, newPurchase)

                    if (newPurchase || oldStatus != licenceStatus || addOnFeatures.size > oldFeatures.size) {
                        (activity as? BillingListener)?.onLicenceStatusSet(prettyPrintStatus(activity))
                    }
                }
                return licenceStatus != null || addOnFeatures.isNotEmpty()
            }

            override fun onPurchaseCanceled() {
                log().i("onPurchasesUpdated() - user cancelled the purchase flow - skipping")
                (activity as? ContribInfoDialogActivity)?.onPurchaseCancelled()
            }

            override fun onPurchaseFailed(resultCode: Int) {
                log().w("onPurchasesUpdated() got unknown resultCode: %s", resultCode)
                (activity as? ContribInfoDialogActivity)?.onPurchaseFailed(resultCode)
            }
        }
        val skuDetailsResponseListener = if (query) SkuDetailsResponseListener { result: BillingResult, skuDetailsList: List<SkuDetails>? ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK && skuDetailsList != null) {
                storeSkuDetails(skuDetailsList)
            } else {
                log().w("skuDetails response %d", result.responseCode)
            }
        } else null
        return BillingManagerPlay(activity, billingUpdatesListener, skuDetailsResponseListener)
    }

    @VisibleForTesting
    fun findHighestValidPurchase(inventory: List<Purchase>) = inventory.mapNotNull { purchase -> extractLicenceStatusFromSku(purchase.sku)?.let { Pair(purchase, it) } }
            .maxByOrNull { pair -> pair.second }?.first

    private fun registerInventory(inventory: List<Purchase>, newPurchase: Boolean) {
        inventory.forEach { purchase: Purchase -> log().i("%s (acknowledged %b)", purchase.sku, purchase.isAcknowledged) }
        findHighestValidPurchase(inventory)?.let {
            if (it.purchaseState == Purchase.PurchaseState.PURCHASED) {
                handlePurchaseForLicence(it.sku, it.orderId)
            } else {
                CrashHandler.reportWithTag(String.format("Found purchase in state %s", it.purchaseState), TAG)
            }
        } ?: run {
            if (!newPurchase) {
                maybeCancel()
            }
        }
        handlePurchaseForAddOns(inventory.map { Pair(Licence.parseFeature(it.sku), it.orderId) }, newPurchase)
        licenseStatusPrefs.commit()
    }

    private fun handlePurchaseForAddOns(features: List<Pair<AddOnPackage?, String>>, newPurchase: Boolean) {
        val newFeatures = features.filter { pair ->
            pair.first?.let {
                persistOrderIdForAddOn(it, pair.second)
                true
            } ?: false
        }.map { it.first!!.feature }
        maybeUpgradeAddonFeatures(newFeatures, newPurchase)
    }

    override fun launchPurchase(aPackage: Package, shouldReplaceExisting: Boolean, billingManager: BillingManager) {
        val sku = getSkuForPackage(aPackage)
        val skuDetails = getSkuDetailsFromPrefs(sku)
                ?: throw IllegalStateException("Could not determine sku details for $sku")
        val oldSku: String?
        if (shouldReplaceExisting) {
            oldSku = currentSubscription
            checkNotNull(oldSku) { "Could not determine current subscription" }
        } else {
            oldSku = null
        }
        (billingManager as BillingManagerPlay).initiatePurchaseFlow(skuDetails, oldSku)
    }
}